### ISSUE TITLE
our assert() is not compatible with the C one, and breaks in edge cases

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cyrus-imapd#4123]
Reported here as an incompatibility with perl 5.36: https://cyrus.topicbox.com/groups/info/T4dbb95a05bf1581e-Mdc484c415098530c67b67f26

We need to either redefine our `assert()` in a way that's compatible with the C one (losing the `{}` and the `if()`), or stop trying to use it as a general replacement for the C one.

**Comment by elliefm** [comment]
In reality, the Cyrus one has some fundamentally different assumptions from the standard one, so the right thing to do is probably rename the Cyrus one.

(We could also fix the definition at the same time, but doing so wouldn't change the underlying different assumptions.)

**Comment by citrus-it** [comment]
> so the right thing to do is probably rename the Cyrus one.

The problem that I saw when testing with perl 5.36 is that, for example, `managesieve.xs` includes `perl.h`, which itself includes `<assert.h>`. The include path has the cyrus `lib/` early, and as a result the _wrong_ assert.h is pulled in, mixing OS and cyrus headers.

```
In file included from /usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/perl.h:3855,
                 from managesieve.xs:46:
/usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/sv_inline.h: In function 'Perl_newSV_type':
../../../lib/assert.h:47:25: error: expected expression before '{' token
   47 | #define assert(ex)      {if (!(ex))assertionfailed(__FILE__, __LINE__, #ex);}
      |                         ^
/usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/handy.h:2787:28: note: in expansion of macro 'assert'
 2787 | #define perl_assert_ptr(p) assert( ((void*)(p)) != 0 )
      |                            ^~~~~~
/usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/handy.h:2792:47: note: in expansion of macro 'perl_assert_ptr'
 2792 | #define Zero(d,n,t)     (MEM_WRAP_CHECK_(n,t) perl_assert_ptr(d), (void)memzero((char*)(d), (n) * sizeof(t)))
      |                                               ^~~~~~~~~~~~~~~
/usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/sv_inline.h:468:13: note: in expansion of macro 'Zero'
  468 |             Zero(new_body, type_details->body_size, char);
      |             ^~~~
gmake: *** [Makefile:346: managesieve.o] Error 1
```

Renaming the `assert.h` header is probably also necessary for a full fix here.

**Comment by elliefm** [comment]
see also #3044
